opm/material/densead/FastSmallVector.hpp

```diff
@@ -0,0 +1,164 @@
+#include <array>
+#include <cstring>
+
+namespace Opm {
+namespace DenseAd {
+
+template <typename ValueType, unsigned N>
+class FastSmallVector
+{
+public:
+
+    FastSmallVector();
+    explicit FastSmallVector(const size_t num_elem);
+    FastSmallVector(const size_t num_elem, const ValueType value);
+    FastSmallVector(const FastSmallVector& other);
+    FastSmallVector(FastSmallVector&& other);
+    ~FastSmallVector();
+
+    FastSmallVector& operator=(const FastSmallVector& other);
+    FastSmallVector& operator=(FastSmallVector&& other);
+
+    ValueType& operator[](size_t index);
+    const ValueType& operator[](size_t index) const;
+    size_t size() const;
+
+private:
+    std::array<ValueType, N> small_buffer_;
+    std::size_t size_;
+    ValueType* data_ptr_;
+};
+
+template <typename ValueType, unsigned N>
+FastSmallVector<ValueType, N>::
```

Comment:
while having the function definitions of templates within the same file is better than the `_impl.hpp` pattern, please adhere to everything else in opm-material and move this into the class body.

Reply:
I put them outside is because I was thinking to put them in a separate file. I like a small class  body so I can understand/check this class in a quick way. 

Anyway, I can move this into the class body.

Reply:
> Anyway, I can move this into the class body.

please do.